packages/store/ts/config/experimental/resolveConfig.ts

```diff
@@ -0,0 +1,168 @@
+import { mapObject } from "@latticexyz/common/utils";
+import { resourceToHex } from "@latticexyz/common";
+
+export type ResolvedStoreConfig<TStoreConfig extends StoreConfig> = TStoreConfig & {
+  /** @deprecated Note: this property is experimental and expected to change */
+  _resolved: {
+    tables: {
+      [TableKey in keyof TStoreConfig["tables"] & string]: ResolvedTableConfig<
+        TStoreConfig["tables"][TableKey],
+        TStoreConfig["userTypes"],
+        keyof TStoreConfig["enums"] & string,
+        TStoreConfig["namespace"],
+        TableKey
+      >;
+    };
+  };
+};
+
+export type ResolvedTableConfig<
+  TTableConfig extends TableConfig,
+  TUserTypes extends UserTypesConfig["userTypes"],
+  TEnumNames extends StringForUnion,
+  TNamespace extends string = string,
+  TName extends string = string
+> = Omit<TTableConfig, "keySchema" | "valueSchema"> & {
+  keySchema: ResolvedKeySchema<TTableConfig["keySchema"], TUserTypes, TEnumNames>;
+  valueSchema: ResolvedValueSchema<TTableConfig["valueSchema"], TUserTypes, TEnumNames>;
+  namespace: TNamespace;
+  name: TName;
+  tableId: `0x${string}`;
+};
+
+export type ResolvedKeySchema<
+  TKeySchema extends TableConfig["keySchema"],
+  TUserTypes extends UserTypesConfig["userTypes"],
+  TEnumNames extends StringForUnion
+> = ResolvedSchema<TKeySchema, TUserTypes, TEnumNames>;
+
+export type ResolvedValueSchema<
+  TValueSchema extends TableConfig["valueSchema"],
+  TUserTypes extends UserTypesConfig["userTypes"],
+  TEnumNames extends StringForUnion
+> = ResolvedSchema<Exclude<TValueSchema, string>, TUserTypes, TEnumNames>;
+
+export type ResolvedSchema<
+  TSchema extends Exclude<TableConfig["keySchema"] | TableConfig["valueSchema"], string>,
+  TUserTypes extends UserTypesConfig["userTypes"],
+  TEnumNames extends StringForUnion
+> = {
+  [key in keyof TSchema]: {
+    type: TSchema[key] extends keyof TUserTypes
+      ? TUserTypes[TSchema[key]] extends UserType
+        ? // Note: we mistakenly named the plain ABI type "internalType",
+          // while in Solidity ABIs the plain ABI type is called "type" and
+          // and the custom type "internalType". We're planning to
+          // change our version and align with Solidity ABIs going forward.
+          TUserTypes[TSchema[key]]["internalType"]
+        : never
+      : TSchema[key] extends TEnumNames
+      ? "uint8"
+      : TSchema[key];
+    internalType: TSchema[key];
+  };
+};
+
+export function resolveConfig<TStoreConfig extends StoreConfig>(
+  config: TStoreConfig
+): ResolvedStoreConfig<TStoreConfig> {
+  const resolvedTables: Record<string, ReturnType<typeof resolveTable>> = {};
+
+  for (const key of Object.keys(config.tables)) {
+    resolvedTables[key] = resolveTable(
+      config.tables[key],
+      config.userTypes,
+      Object.keys(config.enums),
+      config.namespace,
+      key
+    ) as ReturnType<typeof resolveTable>;
+  }
+
+  return {
+    ...config,
```

Comment:
since this function isn't being called by default as part of `mudConfig`, we could just not include `config` here since we'd have a reference to it already being part of the input and only return the `resolved` config (and then no need to worry about naming of `_resolved`)

Reply:
agreed, updated